doc/wasora.1
.\" Automatically generated by Pandoc 2.9.2
.\"
.TH "WASORA" "1" "" "Wasora User Manual" ""
.hy
.SH NAME
.PP
wasora - wasora\[cq]s an advanced suite for optimization & reactor
analysis
.SH SYNOPSIS
.PP
wasora [\f[I]options\f[R]] input-file
[\f[I]optional_extra_arguments\f[R]]\&...
.SH DESCRIPTION
.PP
Wasora (https://www.seamplex.com/wasora) is a (relatively) high-level
tool to perform mathematical computations.
It also provides a framework which other particular computational codes
can use.
It is a free computational tool designed to aid a cognizant
expert\[em]i.e.\ you, whether an engineer, scientist, technician, geek,
etc.\[em]to analyze complex systems by solving mathematical problems by
means of a high-level plain-text input file containing
.IP \[bu] 2
algebraic expressions,
.IP \[bu] 2
data for function interpolation,
.IP \[bu] 2
differential equations, and
.IP \[bu] 2
output instructions
.PP
among other facilities.
.PP
For example, the famous chaotic Lorenz\[cq] dynamical
system (http://en.wikipedia.org/wiki/Lorenz_system)\[em]the one of the
butterfly\[em]whose differential equations are
.PP
.RS
\f[I]x\[u0307]\f[R]\[u2004]=\[u2004]\f[I]\[*s]\f[R]\[u2005]\[md]\[u2005](\f[I]y\f[R]\[u2005]\[mi]\[u2005]\f[I]x\f[R])
.RE
.RS
\f[I]y\[u0307]\f[R]\[u2004]=\[u2004]\f[I]x\f[R]\[u2005]\[md]\[u2005](\f[I]r\f[R]\[u2005]\[mi]\[u2005]\f[I]z\f[R])\[u2005]\[mi]\[u2005]\f[I]y\f[R]
.RE
.RS
\f[I]z\[u0307]\f[R]\[u2004]=\[u2004]\f[I]x\f[R]\f[I]y\f[R]\[u2005]\[mi]\[u2005]\f[I]b\f[R]\f[I]z\f[R]
.RE
.PP
where \f[I]\[*s]\f[R]\[u2004]=\[u2004]10,
\f[I]b\f[R]\[u2004]=\[u2004]8/3 and \f[I]r\f[R]\[u2004]=\[u2004]28 are
the classical parameters that generate the butterfly as presented by
Edward Lorenz back in his seminal 1963 paper Deterministic non-periodic
flow (http://journals.ametsoc.org/doi/abs/10.1175/1520-0469%281963%29020%3C0130%3ADNF%3E2.0.CO%3B2),
can be solved with wasora by writing the equations in the input file as
naturally as possible, as illustrated in the input file that follows:
.IP
.nf
\f[C]
# lorenz\[cq] seminal dynamical system
PHASE_SPACE x y z
end_time = 40

CONST sigma r b
sigma = 10            # parameters
r = 28
b = 8/3

x_0 = -11             # initial conditions
y_0 = -16
z_0 = 22.5

# the dynamical system
x_dot .= sigma*(y - x)
y_dot .= x*(r - z) - y
z_dot .= x*y - b*z

PRINT t x y z HEADER
\f[R]
.fi
.PP
Following the UNIX
Philosophy (https://en.wikipedia.org/wiki/Unix_philosophy), wasora\[cq]s
output can be piped for example to Gnuplot (http://gnuplot.info/) in
order to obtain a beautiful figure:
.IP
.nf
\f[C]
wasora lorenz.was | gnuplot -e \[dq]splot \[aq]-\[aq] u 2:3:4 w l ti \[aq]\[aq]\[dq]
\f[R]
.fi
.PP
[IMAGE: The Lorenz attractor computed by wasora]
.PP
At a first glance, wasora may look as another high-level interpreted
programming language, but\[em]hopefully\[em]it is not: wasora should be
seen as a
syntactically-sweetened (http://en.wikipedia.org/wiki/Syntactic_sugar)
way to ask a computer to perform a certain mathematical computation
(which is what computers do).
.SH OPTIONS
.TP
\f[B]\f[CB]-d\f[B]\f[R] or \f[B]\f[CB]--debug\f[B]\f[R]
Start in debug mode
.TP
\f[B]\f[CB]--node-debug\f[B]\f[R]
Ignore standard input, avoid debug mode
.TP
\f[B]\f[CB]-l\f[B]\f[R] or \f[B]\f[CB]--list\f[B]\f[R]
List defined symbols and exit
.TP
\f[B]\f[CB]-h\f[B]\f[R] or \f[B]\f[CB]--help\f[B]\f[R]
Display this help and exit
.TP
\f[B]\f[CB]-i\f[B]\f[R] or \f[B]\f[CB]--info\f[B]\f[R]
Display detailed code information and exit
.TP
\f[B]\f[CB]-v\f[B]\f[R] or \f[B]\f[CB]--version\f[B]\f[R]
Display version information and exit
.PP
Input file instructions are read from standard input if a dash
\f[C]-\f[R] is passed as \f[C]input-file\f[R].
.PP
Wasora accepts \f[I]optional extra arguments\f[R] which are then
verbatimly replaced in the input file as \f[C]$1\f[R], \f[C]$2\f[R], and
so on.
So for example if an input file has a line like this
.IP
.nf
\f[C]
PRINT \[dq]$1 squared is\[dq] ($1)\[ha]2
\f[R]
.fi
.PP
Then two different meshes called \f[C]one.msh\f[R] and \f[C]two.msh\f[R]
can successively be used in two runs with the same input file by calling
Fino as
.IP
.nf
\f[C]
$ wasora input.was 2
2 squared is    4
$ wasora input.was 3
3 squared is    9
$ 
\f[R]
.fi
.SH REFERENCE
.SS Keywords
.SS =
.RS
.PP
Assign an expression to a variable, a vector or a matrix.
.RE
.IP
.nf
\f[C]
<var>[ [<expr_tmin>, <expr_tmax>] | 
<expr_t> ] = <expr> <vector>(<expr_i>)[<expr_i_min, expr_i_max>] [ [<expr_tmin>, <expr_tmax>] | 
<expr_t> ] = <expr> <matrix>(<expr_i>,<expr_j>)[<expr_i_min, expr_i_max; expr_j_min, expr_j_max>] [ [<expr_tmin>, <expr_tmax>] | 
<expr_t> ] = <expr>
\f[R]
.fi
.SS _.=
.RS
.PP
Add an equation to the DAE system to be solved in the phase space
spanned by \f[C]PHASE_SPACE\f[R].
.RE
.IP
.nf
\f[C]
{ 0[(i[,j]][<imin:imax[;jmin:jmax]>] | <expr1> } .= <expr2>
\f[R]
.fi
.SS ABORT
.RS
.PP
Catastrophically abort the execution and quit wasora.
.RE
.IP
.nf
\f[C]
ABORT
\f[R]
.fi
.PP
Whenever the instruction \f[C]ABORT\f[R] is executed, wasora quits
without closing files or unlocking shared memory objects.
The objective of this instruction is, as illustrated in the examples,
either to debug complex input files and check the values of certain
variables or to conditionally abort the execution using \f[C]IF\f[R]
clauses.
.SS ALIAS
.RS
.PP
Define a scalar alias of an already-defined indentifier.
.RE
.IP
.nf
\f[C]
ALIAS { <new_var_name> IS <existing_object> | <existing_object> AS <new_name> }
\f[R]
.fi
.PP
The existing object can be a variable, a vector element or a matrix
element.
In the first case, the name of the variable should be given as the
existing object.
In the second case, to alias the second element of vector \f[C]v\f[R] to
the new name \f[C]new\f[R], \f[C]v(2)\f[R] should be given as the
existing object.
In the third case, to alias second element (2,3) of matrix \f[C]M\f[R]
to the new name \f[C]new\f[R], \f[C]M(2,3)\f[R] should be given as the
existing object.
.SS CALL
.RS
.PP
Call a previously dynamically-loaded user-provided routine.
.RE
.IP
.nf
\f[C]
CALL <name> [ expr_1 expr_2 ... expr_n ]
\f[R]
.fi
.SS CLOSE
.RS
.PP
Explicitly close an already-\f[C]OPEN\f[R]ed file.
.RE
.IP
.nf
\f[C]
CLOSE
\f[R]
.fi
.SS CONST
.RS
.PP
Mark a scalar variable, vector or matrix as a constant.
.RE
.IP
.nf
\f[C]
CONST name_1 [ <name_2> ] ... [ <name_n> ]
\f[R]
.fi
.SS DEFAULT_ARGUMENT_VALUE
.RS
.PP
Give a default value for an optional commandline argument.
.RE
.IP
.nf
\f[C]
DEFAULT_ARGUMENT_VALUE <constant> <string>
\f[R]
.fi
.PP
If a \f[C]$n\f[R] construction is found in the input file but the
commandline argument was not given, the default behavior is to fail
complaining that an extra argument has to be given in the commandline.
With this keyword, a default value can be assigned if no argument is
given, thus avoiding the failure and making the argument optional.
.SS DIFFERENTIAL
.RS
.PP
Explicitly mark variables, vectors or matrices as \[lq]differential\[rq]
to compute intial conditions of DAE systems.
.RE
.IP
.nf
\f[C]
DIFFERENTIAL { <var_1> <var_2> ... | <vector_1> <vector_2> ... | <matrix_1> <matrix_2> ... }
\f[R]
.fi
.SS DO_NOT_EVALUATE_AT_PARSE_TIME
.RS
.PP
Ask wasora not to evaluate assignments at parse time.
.RE
.IP
.nf
\f[C]
DO_NOT_EVALUATE_AT_PARSE_TIME
\f[R]
.fi
.SS FILE
.RS
.PP
Define a file, either as input or as output, for further usage.
.RE
.IP
.nf
\f[C]
< FILE | OUTPUT_FILE | INPUT_FILE > <name> <printf_format> [ expr_1 expr_2 ... expr_n ] [ INPUT | OUTPUT | MODE <fopen_mode> ] [ OPEN | DO_NOT_OPEN ]
\f[R]
.fi
.SS FIT
.RS
.PP
Fit a function of one or more arguments to a set of pointwise-defined
data.
.RE
.IP
.nf
\f[C]
FIT <function_to_be_fitted> TO <function_with_data> VIA <var_1> <var_2> ... <var_n>
 [ GRADIENT <expr_1> <expr_2> ... <expr_n> ]
 [ RANGE_MIN <expr_1> <expr_2> ... <expr_j> ]
 [ RANGE_MAX <expr_1> <expr_2> ... <expr_n> ]
 [ DELTAEPSREL <expr> ] [ DELTAEPSABS <expr> ] [ MAX_ITER <expr> ]
 [ VERBOSE ] [ RERUN | DO_NOT_RERUN ]
\f[R]
.fi
.PP
The function with the data has to be point-wise defined (i.e.\ a
\f[C]FUNCTION\f[R] read from a file with inline \f[C]DATA\f[R]).
The function to be fitted has to be parametrized with at least one of
the variables provided after the \f[C]VIA\f[R] keyword.
Only the names of the functions have to be given, not the arguments.
Both functions have to have the same number of arguments.
The initial guess of the solution is given by the initial value of the
variables listed in the \f[C]VIA\f[R] keyword.
Analytical expressions for the gradient of the function to be fitted
with respect to the parameters to be fitted can be optionally given with
the \f[C]GRADIENT\f[R] keyword.
If none is provided, the gradient will be computed numerically using
finite differences.
A range over which the residuals are to be minimized can be given with
\f[C]RANGE_MIN\f[R] and \f[C]RANGE_MAX\f[R].
The expressions give the range of the arguments of the functions, not of
the parameters.
For multidimensional fits, the range is an hypercube.
If no range is given, all the definition points of the function with the
data are used for the fit.
Convergence can be controlled by giving the relative and absolute
tolreances with \f[C]DELTAEPSREL\f[R] (default \f[C]1e-4\f[R]) and
\f[C]DELTAEPSABS\f[R] (default \f[C]1e-6\f[R]), and with the maximum
number of iterations \f[C]MAX_ITER\f[R] (default 100).
If the optional keyword \f[C]VERBOSE\f[R] is given, some data of the
intermediate steps is written in the standard output.
The combination of arguments that minimize the function are computed and
stored in the variables.
So if \f[C]f(x,y)\f[R] is to be minimized, after a \f[C]MINIMIZE f\f[R]
both \f[C]x\f[R] and \f[C]y\f[R] would have the appropriate values.
The details of the method used can be found in GSL\[cq]s
documentation (https://www.gnu.org/software/gsl/doc/html/multimin.html).
Some of them use derivatives and some of them do not.
Default method is \f[C]gsl_multimin_fminimizer_nmsimplex2\f[R], which
does not need derivatives.
.SS FUNCTION
.RS
.PP
Define a function of one or more variables.
.RE
.IP
.nf
\f[C]
FUNCTION <name>(<var_1>[,var2,...,var_n]) { [ = <expr> | FILE_PATH <file_path> | ROUTINE <name> | | MESH <name> { DATA <new_vector_name> | VECTOR <existing_vector_name> } { NODES | CELLS } | [ VECTOR_DATA <vector_1> <vector_2> ... <vector_n> <vector_n+1> ] } [COLUMNS <expr_1> <expr_2> ... <expr_n> <expr_n+1> ] [ INTERPOLATION { linear | polynomial | spline | spline_periodic | akima | akima_periodic | steffen | nearest | shepard | shepard_kd | bilinear } ] [ INTERPOLATION_THRESHOLD <expr> ] [ SHEPARD_RADIUS <expr> ] [ SHEPARD_EXPONENT <expr> ] [ SIZES <expr_1> <expr_2> ... <expr_n> ] [ X_INCREASES_FIRST <expr> ] [ DATA <num_1> <num_2> ... <num_N> ]
\f[R]
.fi
.PP
The number of variables \f[I]n\f[R] is given by the number of arguments
given between parenthesis after the function name.
The arguments are defined as new variables if they had not been already
defined as variables.
If the function is given as an algebraic expression, the short-hand
operator \f[C]:=\f[R] can be used.
That is to say, \f[C]FUNCTION f(x) = x\[ha]2\f[R] is equivalent to
\f[C]f(x) := x\[ha]2\f[R].
If a \f[C]FILE_PATH\f[R] is given, an ASCII file containing at least
\f[I]n\f[R]\[u2005]+\[u2005]1 columns is expected.
By default, the first \f[I]n\f[R] columns are the values of the
arguments and the last column is the value of the function at those
points.
The order of the columns can be changed with the keyword
\f[C]COLUMNS\f[R], which expects \f[I]n\f[R]\[u2005]+\[u2005]1
expressions corresponding to the column numbers.
A function of type \f[C]ROUTINE\f[R] calls an already-defined
user-provided routine using the \f[C]CALL\f[R] keyword and passes the
values of the variables in each required evaluation as a
\f[C]double *\f[R] argument.
If \f[C]MESH\f[R] is given, the definition points are the nodes or the
cells of the mesh.
The function arguments should be (\f[I]x\f[R]),
(\f[I]x\f[R],\[u2006]\f[I]y\f[R]) or
(\f[I]x\f[R],\[u2006]\f[I]y\f[R],\[u2006]\f[I]z\f[R]) matching the
dimension the mesh.
If the keyword \f[C]DATA\f[R] is used, a new empty vector of the
appropriate size is defined.
The elements of this new vector can be assigned to the values of the
function at the \f[I]i\f[R]-th node or cell.
If the keyword \f[C]VECTOR\f[R] is used, the values of the dependent
variable are taken to be the values of the already-existing vector.
Note that this vector should have the size of the number of nodes or
cells the mesh has, depending on whether \f[C]NODES\f[R] or
\f[C]CELLS\f[R] is given.
If \f[C]VECTOR_DATA\f[R] is given, a set of
\f[I]n\f[R]\[u2005]+\[u2005]1 vectors of the same size is expected.
The first \f[I]n\f[R]\[u2005]+\[u2005]1 correspond to the arguments and
the last one is the function value.
Interpolation schemes can be given for either one or multi-dimensional
functions with \f[C]INTERPOLATION\f[R].
Available schemes for \f[I]n\f[R]\[u2004]=\[u2004]1 are:
.IP \[bu] 2
linear
.IP \[bu] 2
polynomial, the grade is equal to the number of data minus one
.IP \[bu] 2
spline, cubic (needs at least 3 points)
.IP \[bu] 2
spline_periodic
.IP \[bu] 2
akima (needs at least 5 points)
.IP \[bu] 2
akima_periodic (needs at least 5 points)
.IP \[bu] 2
steffen, always-monotonic splines-like (available only with GSL >= 2.0)
.PP
Default interpolation scheme for one-dimensional functions is
\f[C](*gsl_interp_linear)\f[R].
.PP
Available schemes for \f[I]n\f[R]\[u2004]>\[u2004]1 are:
.IP \[bu] 2
nearest, \f[I]f\f[R](\f[I]x\[u20D7]\f[R]) is equal to the value of the
closest definition point
.IP \[bu] 2
shepard, inverse distance weighted average definition
points (https://en.wikipedia.org/wiki/Inverse_distance_weighting) (might
lead to inefficient evaluation)
.IP \[bu] 2
shepard_kd, average of definition points within a
kd-tree (https://en.wikipedia.org/wiki/Inverse_distance_weighting#Modified_Shepard's_method)
(more efficient evaluation provided \f[C]SHEPARD_RADIUS\f[R] is set to a
proper value)
.IP \[bu] 2
bilinear, only available if the definition points configure an
structured hypercube-like grid.
If \f[I]n\f[R]\[u2004]>\[u2004]3, \f[C]SIZES\f[R] should be given.
.PP
For \f[I]n\f[R]\[u2004]>\[u2004]1, if the euclidean distance between the
arguments and the definition points is smaller than
\f[C]INTERPOLATION_THRESHOLD\f[R], the definition point is returned and
no interpolation is performed.
Default value is square root of \f[C]9.5367431640625e-07\f[R].
The initial radius of points to take into account in
\f[C]shepard_kd\f[R] is given by \f[C]SHEPARD_RADIUS\f[R].

NEWS
See <https://www.seamplex.com/wasora/> for news.


src/wasora.c
/*------------ -------------- -------- --- ----- ---   --       -            -
 *  wasora main function
 *
 *  Copyright (C) 2009--2020 jeremy theler
 *
 *  This file is part of wasora.
 *
 *  wasora is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  wasora is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with wasora.  If not, see <http://www.gnu.org/licenses/>.
 *------------------- ------------  ----    --------  --     -       -         -
 */

#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <string.h>
#include <unistd.h>
#include <getopt.h>
#include <libgen.h>

#include <gsl/gsl_errno.h>

#ifndef _WASORA_H_
#include "wasora.h"
#endif

int main(int argc, char **argv) {

  int i, optc;
  int option_index = 0;
  int show_help = 0;
  int show_version = 0;
  
  const struct option longopts[] = {
    { "help",     no_argument,       NULL, 'h'},
    { "version",  no_argument,       NULL, 'v'},
    { "info",     no_argument,       NULL, 'i'},
    { "plugin",   required_argument, NULL, 'P'},
    { "no-debug", no_argument,       NULL, 'n'},
    { "debug",    no_argument,       NULL, 'd'},
    { "list",     no_argument,       NULL, 'l'},
    { NULL, 0, NULL, 0 }
  };  

#if HARDCODEDPLUGIN
  if (wasora_load_hardcoded_plugin() != WASORA_RUNTIME_OK) {
    wasora_runtime_error();
    exit(EXIT_FAILURE);
  }
#endif

  // antes de que getopt nos rompa cosas, nos copiamos los argumentos
  wasora.argc_orig = argc;
  wasora.argv_orig = malloc((argc+1) * sizeof(char *));
  for (i = 0; i < argc; i++) {
    wasora.argv_orig[i] = strdup(argv[i]);
  }
  wasora.argv_orig[i] = NULL;

  // que no se queje si una opcion no es valida para wasora,
  // puede ser valida para algun plugin o algo
  opterr = 0;
  while ((optc = getopt_long_only(argc, argv, "hviP:dl", longopts, &option_index)) != -1) {
    switch (optc) {
      case 'h':
        show_help = 1;
        break;
      case 'v':
        show_version = VERSION_COPYRIGHT;
        break;
      case 'i':
        show_version = VERSION_INFO;
        break;
      case 'p':
        if (wasora_load_plugin(optarg) != WASORA_RUNTIME_OK) {
          wasora_push_error_message("loading plugin %s", optarg);
          wasora_pop_errors();
          exit(EXIT_FAILURE);
        }
        break;
      case 'n':
        wasora.mode = mode_ignore_debug;
        break;
      case 'd':
        wasora.mode = mode_debug;
        break;
      case 'l':
        wasora.mode = mode_list_vars;
        break;
      case '?':
        break;
      default:
        break;
    }
  }
  
  if (show_help) {
    wasora_show_help(argv[0]);
    wasora_finalize();
    exit(EXIT_SUCCESS);
  } else if (show_version && optind == argc) {
    wasora_show_version(show_version);
    wasora_finalize();
    exit(EXIT_SUCCESS);
  } else if (wasora.mode && optind == argc) {
    wasora_call(wasora_init_before_parser());
    wasora_list_symbols();
    wasora_finalize();
    exit(EXIT_SUCCESS);
  } else if (optind == argc) {
    wasora_show_version(0);
    wasora_finalize();
    exit(EXIT_SUCCESS);
  }
   
  // nos arcordamos de los argumentos de la linea de comandos y del indice 
  // desde donde empieza el input (y despues los argumentos a reemplazar)
  // alocamos y copiamos oprque despues los podemos toquetear con DEFAULT_ARGUMENT_VALUE
  wasora.argc = argc;
  wasora.argv = malloc(sizeof(char *)*argc);
  memcpy(wasora.argv, argv, sizeof(char *)*argc);
  wasora.optind = optind;
  
  // nos acordamos del directorio del input para intentar
  // otras alternativas al abrir files
  wasora.main_input_dirname_aux = strdup(argv[optind]);
  wasora.main_input_dirname = dirname(wasora.main_input_dirname_aux);

  // sacamos el handler de la gsl y metemos el nuestro
  gsl_set_error_handler(wasora_gsl_handler);

  // instalamos signal handlers para int, kill y chld
  signal(SIGINT, wasora_signal_handler);
  signal(SIGTERM, wasora_signal_handler);
//  signal(SIGCHLD, wasora_signal_handler);

  if (wasora_parse_main_input_file(argv[optind]) != 0) {
    wasora_pop_errors();
// si hubo un error de parseo no hacemos finalize porque puede haber algo roto
//    wasora_finalize();
    exit(WASORA_PARSER_ERROR);
  }
 
  // vemos como tenemos que correr
  if (show_version) {
    wasora_show_version(1);
    exit(EXIT_SUCCESS);
  } else if (wasora.mode == mode_list_vars) {
    wasora_list_symbols();
    i = WASORA_RUNTIME_OK;
  } else if (wasora.parametric_mode) {
    i = wasora_parametric_run();
  } else if (wasora.min_mode) {
    i = wasora_min_run();
  } else if (wasora.fit_mode) {
    i = wasora_fit_run();
  } else {
    i = wasora_standard_run();
  }
  
  if (i != WASORA_RUNTIME_OK) {
//    wasora_pop_errors();
    wasora_runtime_error();
  }

  // bye bye
  wasora_finalize();
  
  return WASORA_RUNTIME_OK;

}



int wasora_standard_run(void) {

#ifdef HAVE_IDA
  int err;
  double ida_step_dt, ida_step_t_old, ida_step_t_new, t_old;
#endif

  wasora_call(wasora_init_before_run());
  
  // calculo estatico
  wasora_value(wasora_special_var(in_static)) = 1;
  wasora_value(wasora_special_var(in_static_first)) = 1;
  while (!wasora_value(wasora_special_var(done)) && !wasora_value(wasora_special_var(done_static)) && (int)(wasora_var(wasora_special_var(step_static))) < rint(wasora_var(wasora_special_var(static_steps)))) {

    wasora_debug();

    wasora_var(wasora_special_var(step_static)) = wasora_var(wasora_special_var(step_static)) + 1;
    if ((int)wasora_var(wasora_special_var(step_static)) != 1 && (int)wasora_var(wasora_special_var(step_static)) == (int)wasora_var(wasora_special_var(static_steps))) {
      wasora_value(wasora_special_var(in_static_last)) = 1;
    }

    if (wasora_dae.daes == NULL) {
      
      wasora_call(wasora_step(STEP_ALL));
      
    } else {
      
      wasora_call(wasora_step(STEP_BEFORE_DAE));
      if (wasora_dae.system == NULL) {
        wasora_call(wasora_dae_init());
      }
      wasora_call(wasora_dae_ic());
      wasora_call(wasora_step(STEP_AFTER_DAE));
      
    }
    wasora_value(wasora_special_var(in_static_first)) = 0;
  }
  wasora_value(wasora_special_var(in_static)) = 0;
  wasora_value(wasora_special_var(in_static_last)) = 0;
  wasora_value(wasora_special_var(step_static)) = 0;

  
  
#ifdef HAVE_IDA
  ida_step_dt = INFTY;
#endif
  
  // si hay realtime, inicializamos despues del calculo estationario por si tuvimos
  // que esperar el semaforo de alguien o algo por el estilo
  if (wasora_var(wasora.special_vars.realtime_scale) != 0) {
    wasora_init_realtime();
  }
  
  // loop  transitorio (si es necesario)
  wasora_value(wasora_special_var(in_transient)) = 1;
  wasora_value(wasora_special_var(in_transient_first)) = 1;
  while (wasora_var(wasora_special_var(done)) == 0) {

    wasora_debug();
    
    wasora_value(wasora_special_var(step_transient)) = wasora_value(wasora_special_var(step_transient)) + 1;

    if (wasora_dae.dimension == 0) {
    
      wasora.next_time = wasora_var(wasora_special_var(time)) + wasora_var(wasora_special_var(dt));
      // si nos dieron un time path, vemos que no nos hayamos pasado
      if (wasora.time_path != NULL && wasora.current_time_path->n_tokens != 0 && wasora.next_time > wasora_evaluate_expression(wasora.current_time_path)) {
        // nos paramos un cachito mas adelante del next time asi ya los
        // assigns toman los nuevos valores de la posible discontinuidad
        // que hay en el punto en cuestion */
        wasora.next_time = wasora_evaluate_expression(wasora.current_time_path) + wasora_var(wasora_special_var(zero));
        wasora.current_time_path++;
      }

      // dormimos si hay que hacer realtime
      if (wasora_var(wasora.special_vars.realtime_scale) != 0) {
        wasora_wait_realtime();
      }

      // actualizamos time, si esta importada despues nos la pisan asi que no pasa nada
      wasora_var(wasora_special_var(dt)) = wasora.next_time - wasora_var(wasora_special_var(time));
      wasora_var(wasora_special_var(time)) = wasora.next_time;
      
      if (wasora_var(wasora_special_var(time)) >= wasora_var(wasora_special_var(end_time))) {
        wasora_value(wasora_special_var(in_transient_last)) = 1;
        wasora_value(wasora_special_var(done_transient)) = 1;
      }
      
      // transient step
      wasora_call(wasora_step(STEP_ALL));
      
    } else {
      
#ifdef HAVE_IDA
        
      wasora_call(wasora_step(STEP_BEFORE_DAE));
      // integration step
      // nos acordamos de cuanto valia el tiempo antes de avanzar un paso
      // para despues saber cuanto vale dt
      t_old = wasora_var(wasora_special_var(time));

      // miramos si hay max_dt
      if (wasora_var(wasora_special_var(max_dt)) != 0) {
        IDASetMaxStep(wasora_dae.system, wasora_var(wasora_special_var(max_dt)));
      }

      // miramos si el dt actual (del paso interno de ida) es mas chiquito que min_dt
      if (ida_step_dt < wasora_var(wasora_special_var(min_dt)) || (wasora_var(wasora_special_var(min_dt)) != 0 && wasora_var(wasora_special_var(time)) == 0)) {

        if (wasora.current_time_path != NULL && wasora.current_time_path->n_tokens != 0) {
          wasora_push_error_message("both min_dt and TIME_PATH given");
          wasora_runtime_error();
          return WASORA_RUNTIME_ERROR;
        }

        // ponemos stop time para que no nos pasemos
        // TODO: ver si no conviene llamar con el flag IDA_NORMAL
        IDASetStopTime(wasora_dae.system, wasora_var(wasora_special_var(time))+wasora_var(wasora_special_var(min_dt)));
        do {
          ida_step_t_old = wasora_var(wasora_special_var(time));
          ida_step_t_new = wasora_var(wasora_special_var(time))+wasora_var(wasora_special_var(min_dt));
          err = IDASolve(wasora_dae.system, wasora_var(wasora_special_var(dt)), &ida_step_t_new, wasora_dae.x, wasora_dae.dxdt, IDA_ONE_STEP);
  //        err = IDASolve(wasora_dae.system, wasora_var(wasora_special_var(dt)), &ida_step_t_new, wasora_dae.x, wasora_dae.dxdt, IDA_NORMAL);
          if (err < 0) {
            wasora_push_error_message("ida returned error code %d", err);
            wasora_runtime_error();
            return WASORA_RUNTIME_ERROR;
          }
          wasora_var(wasora_special_var(time)) = ida_step_t_new;
          ida_step_dt = ida_step_t_new - ida_step_t_old;
        } while (err != IDA_TSTOP_RETURN);

      } else {

        // miramos si hay TIME_PATH
        // nos paramos un cachito mas adelante para q los steps ya hayan actuado
        if (wasora.current_time_path != NULL && wasora.current_time_path->n_tokens != 0) {
          IDASetStopTime(wasora_dae.system, wasora_evaluate_expression(wasora.current_time_path)+wasora_var(wasora_special_var(zero)));
  //        IDASetStopTime(wasora_dae.system, evaluate_expression(wasora.current_time_path)+dt0);
        }

        ida_step_t_old = wasora_var(wasora_special_var(time));
        err = IDASolve(wasora_dae.system, wasora_var(wasora_special_var(end_time)), &wasora_var(wasora_special_var(time)), wasora_dae.x, wasora_dae.dxdt, IDA_ONE_STEP);
        ida_step_dt = wasora_var(wasora_special_var(time)) - ida_step_t_old;

  /*      
        if (wasora.in_static_step == 0 && (ida_step_dt = wasora_var(wasora_special_var(time)) - ida_step_t_old) < wasora_var(wasora_special_var(zero))) {
          wasora_push_error_message("ida returned dt = 0, usually meaning that the equations do not converge to the selected tolerance, try setting a higher error_bound or smoothing discontinuous functions of time");
          wasora_runtine_error();
          return;
        }
  */
        if (err == IDA_SUCCESS) {
          ; // ok!
        } else if (err == IDA_TSTOP_RETURN) {
          ++wasora.current_time_path;
        } else {
          wasora_push_error_message("ida returned error code %d", err);
          return WASORA_RUNTIME_ERROR;
        }
      }

      wasora_var(wasora_special_var(dt)) = wasora_var(wasora_special_var(time)) - t_old;
      if (wasora_var(wasora_special_var(time)) >= wasora_var(wasora_special_var(end_time))) {
        wasora_value(wasora_special_var(in_transient_last)) = 1;
        wasora_value(wasora_special_var(done_transient)) = 1;
      }
      
      wasora_call(wasora_step(STEP_AFTER_DAE));
      
      // dormimos si hay que hacer realtime
      if (wasora_var(wasora.special_vars.realtime_scale) != 0) {
        wasora_wait_realtime();
      }
      
#endif      
    }
    
    // listo! (bug encontrado por ramiro)
    wasora_value(wasora_special_var(in_transient_first)) = 0;

  }
  
  
  return WASORA_RUNTIME_OK;
}



int wasora_step(int whence) {
  instruction_t *first, *last;

  // ponemos done en true si nos pasamos para que este disponible durante el step
  if ((int)(wasora_var(wasora_special_var(in_static)))) {
    if ((int)(wasora_var(wasora_special_var(step_static))) != 1 && (int)(wasora_var(wasora_special_var(step_static))) >= (int)(wasora_var(wasora_special_var(static_steps)))) {
      wasora_var(wasora_special_var(done_static)) = 1;
    }
  }
  if (wasora_var(wasora_special_var(end_time)) == 0) {
    wasora_var(wasora_special_var(done)) = wasora_var(wasora_special_var(done_static));
  } else if (wasora_var(wasora_special_var(time)) >= (wasora_var(wasora_special_var(end_time)) - 1e-1*wasora_var(wasora_special_var(dt)))) {
    wasora_var(wasora_special_var(done)) = 1;
  }  
  
  switch (whence) {
    case STEP_BEFORE_DAE:
      first = wasora.instructions;
      last = wasora_dae.instruction;
    break;
    case STEP_AFTER_DAE:
      first = wasora_dae.instruction;
      last = NULL;
    break;
    default:
      first = wasora.instructions;
      last = NULL;
    break;
  }

  // barremos la lista en los extremos que nos dijeron
  // pero siguiendo la logica de condicionales
  wasora.ip = first;
  while (wasora.ip != last) {
    wasora_call(wasora.ip->routine(wasora.ip->argument));

    if (wasora.next_flow_instruction != NULL) {
      wasora.ip = wasora.next_flow_instruction;
      wasora.next_flow_instruction = NULL;
    } else {
      wasora.ip = wasora.ip->next;
    }
  }


  // volvemos a poner done en true por si algun salame puso done = 0 por alguna otra razon
  if ((int)(wasora_var(wasora_special_var(in_static)))) {
    if ((int)(wasora_var(wasora_special_var(step_static))) >= (int)(wasora_var(wasora_special_var(static_steps)))) {
      wasora_var(wasora_special_var(done_static)) = 1;
    }
  }
  if (wasora_var(wasora_special_var(end_time)) == 0) {
    wasora_var(wasora_special_var(done)) = wasora_var(wasora_special_var(done_static));
  } else if (wasora_var(wasora_special_var(time)) >= (wasora_var(wasora_special_var(end_time)) - 1e-1*wasora_var(wasora_special_var(dt)))) {
    wasora_var(wasora_special_var(done)) = 1;
  }  

  return WASORA_RUNTIME_OK;

}


int wasora_instruction_abort(void *arg) {
  exit(1);
}
